common/src/primitives/per_thousand.rs

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use serialization::{Decode, Encode};
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Debug, Encode, Decode)]
```

Comment:
Note about type-safety: Whenever Decode exists, you basically have a de-facto method to circumvent whatever type-safety rules you created. Meaning: If you think now that `PerThousand` can never have a `value >= 1000`, think again. This can now be circumvented through `Decode`.

In other words: Never rely on type-safety alone if you'll allow Decode.

You can see this problem being solved in the multisig implementation I created a monto or two ago, for example. It's the same issue.

IF you want to do this, then you have to implement `Decode` yourself and make decoding a `value >= 1000` an error.

Reply:
will do, thanks for pointing out